### PR TITLE
[Variant]   Fuzz testing and benchmarks for vaildation

### LINKED ISSUE #7842: [Variant] Add testing for invalid variants (fuzz testing??)
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
- @scovich  added some great changes in https://github.com/apache/arrow-rs/pull/7807 to improve the validation APIs in variant. 

However, we don't really have much coverage of the validation code (specifically validation on input errors).       

**Describe the solution you'd like**

I would like additional coverage of the variant error cases

**Describe alternatives you've considered**

I do think trying to cook up examples of all possible malformed variants is probably not a great use of time, but writing a fuzz tester might be

> Maybe we could create a fuzz tester that created a random variant using the builder, and then randomly changed a few bytes around. -- Then if validation succeeded we would walk over the entire variant and verify that the infallable APIs didn't panic

I think fuzzing would help a lot, especially with AFL

_Originally posted by @friendlymatthew in https://github.com/apache/arrow-rs/issues/7807#issuecomment-3023810475_
**Additional context**

### PR DESCRIPTION
**Pull request by carpecodeum**
# Which issue does this PR close?

Closes #7842 (Add testing for invalid variants)

# Rationale for this change

After adding support for both fallible and infallible access to variants, @alamb  pointed out that there aren't many tests for the validation system itself.
CC - @scovich @friendlymatthew 

# What changes are included in this PR?

This change adds the fuzzing @alamb  requested: it generates valid variants using the builder, randomly corrupts them by flipping bits, then tests both validation paths (if validation passes, make sure access doesn't crash; if it fails, make sure error handling works properly) across many corruption scenarios plus specific malformed test cases.

A huge thank you to @PinkCrow007, @mprammer, @alamb, and the rest of the CMU variant team for their continued support towards this project.

# Are these changes tested?
Yes, passing all the tests currently

We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?

# Are there any user-facing changes?

No, tests are to make sure the validation system works fine

### CODE REVIEW

**Comment by alamb** [comment]
> For the fuzzing though, I'm wondering if we considered using something more robust like [AFL](https://rust-fuzz.github.io/book/afl.html). Afl does coverage guided fuzzing which should be more effective at finding weird edge cases than just flipping bits at random, because it can use code coverage to determine which paths to take when mutating test cases. It also gets us test minimization for free

I recommend we file another ticket to look into a more robust / automated framework for fuzzing. It applies to more than just variant and would make a nice addition to this crate I think (though I also think it will be a non trivial thing to setup, but I haven't actually tried it)

File: parquet-variant/benches/variant_builder.rs [hunk 1]
```diff
@@ -388,6 +388,113 @@ fn bench_object_list_partially_same_schema(c: &mut Criterion) {
     });
 }
 
+// Benchmark validation performance
+fn bench_validation_validated_vs_unvalidated(c: &mut Criterion) {
```
Comment (alamb) [inline parquet-variant/benches/variant_builder.rs:392]: This is a cool idea -- not related to fuzz testing I don't think but a nice addition anyways

File: parquet-variant/tests/variant_interop.rs [hunk 2]
```diff
@@ -275,3 +278,309 @@ fn variant_object_builder() {
...
+    rng: &mut StdRng,
+    mut metadata: Vec<u8>,
+    mut value: Vec<u8>,
+) -> (Vec<u8>, Vec<u8>) {
+    // Randomly decide what to corrupt
+    let corrupt_metadata = rng.random_bool(0.3);
+    let corrupt_value = rng.random_bool(0.7);
+
+    if corrupt_metadata && !metadata.is_empty() {
+        let num_corruptions = rng.random_range(1..=(metadata.len().min(5)));
```
Comment (alamb) [inline parquet-variant/tests/variant_interop.rs]: i wonder if we need to corrupt it with more than one corruption 🤔 

If you put this many corruptions I think it means you'll basically correct the entire thing for small variants (which might be ok

Maybe it should be `max(3)` instead to do up to three corruptions 🤔

**Review by alamb — APPROVED** [review]
Thank you @carpecodeum  -- I think this is a very nice test addition. I left some comments but I also think we could merge it in as is

I also verified this covers many of the edge cases using [`llvm-cov`](https://github.com/taiki-e/cargo-llvm-cov)

```shell
 cargo llvm-cov --html -p parquet-variant --tests -- test_validation_fuzz_integration
```

And then I reviewed the report (attached: [report.zip](https://github.com/user-attachments/files/21045382/report.zip))

And found that many more error paths are covered 👍

**Comment by carpecodeum** [comment]
> Hi, really like the new tests for the malformed variants.
> 
> For the fuzzing though, I'm wondering if we considered using something more robust like [AFL](https://rust-fuzz.github.io/book/afl.html). Afl does coverage guided fuzzing which should be more effective at finding weird edge cases than just flipping bits at random, because it can use code coverage to determine which paths to take when mutating test cases. It also gets us test minimization for free

I do agree! I initially planned to include something like afl, and thought maybe I could extend that functionality in this PR with more direction from you and @alamb, but Andrew states a very good point that we should probably make a new ticket and extend that to the crate. If you guys want, I can make a new PR making that setup

File: parquet-variant/tests/variant_interop.rs (hunk 2 above)
Comment (carpecodeum) [reply parquet-variant/tests/variant_interop.rs]: I changed to exactly 1 corruption is that fine?

**Comment by scovich** [comment]
Aside: The coverage report exposed the fact that `map_try_from_slice_error` is dead code. Maybe we should delete it?

**Comment by alamb** [comment]
> Aside: The coverage report exposed the fact that `map_try_from_slice_error` is dead code. Maybe we should delete it?

I agree
- https://github.com/apache/arrow-rs/pull/7861

**Comment by alamb** [comment]
Thanks @carpecodeum

**Comment by alamb** [comment]
> but Andrew states a very good point that we should probably make a new ticket and extend that to the crate. If you guys want, I can make a new PR making that setup

I think it would be worth a ticket and a quick proof of concept 

I don't have any experience with ths AFL framework and I don't know how well suited it is for runing in CI normally
